core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -821,9 +858,131 @@ object Parser {
+      i = Statics.mix(i, Statics.anyHash(input))
+      i = Statics.mix(i, failedAtOffset)
+      i = Statics.mix(i, Statics.anyHash(expected))
+      Statics.finalizeHash(i, 2)
+    }
+    override def toString(): String = s"Error($failedAtOffset, $expected)"
+    override def equals(x$1: Any): Boolean = {
+      if (x$1.isInstanceOf[Error]) {
+        val other = x$1.asInstanceOf[Error]
+        other.input == input &&
+        other.failedAtOffset == failedAtOffset &&
+        other.expected == expected
+
+      } else false
+    }
+  }
+
+  object ErrorWithInput {
+    def unapply(error: Error): Option[(String, Int, NonEmptyList[Expectation])] =
+      error.input.map(input => (input, error.failedAtOffset, error.expected))
+  }
+
+  object Error extends Serializable {
+    def apply(failedAtOffset: Int, expected: NonEmptyList[Expectation]): Error =
+      new Error(None, failedAtOffset, expected)
+
+    def apply(input: String, failedAtOffset: Int, expected: NonEmptyList[Expectation]): Error =
+      new Error(Some(input), failedAtOffset, expected)
+
+    def unapply(error: Error): Some[(Int, NonEmptyList[Expectation])] =
+      Some((error.failedAtOffset, error.expected))
+
+    implicit val catsShowError: Show[Error] =
+      new Show[Error] {
+        def show(error: Error): String = {
+          error.input match {
+            case Some(input) => {
+              val locationMap = new LocationMap(input)
+              def errorMsg = error.expected.map(e => s"* ${e.show}").toList.mkString("\n")
+
+              locationMap.toCaret(error.failedAtOffset) match {
+                case None => errorMsg
+                case Some(caret) => {
+                  val lines = locationMap.lines
+
+                  val contextSize = 2
+
+                  val start = caret.line - contextSize
+                  val end = caret.line + 1 + contextSize
+
+                  val elipsis = "..."
+
+                  val beforeElipsis =
+                    if (start <= 0) None
+                    else Some(elipsis)
+
+                  val beforeContext =
+                    Some(lines.slice(start, caret.line).mkString("\n")).filter(_.nonEmpty)
+
+                  val line = lines(caret.line)
+
+                  val afterContext: Option[String] =
+                    Some(lines.slice(caret.line + 1, end).mkString("\n")).filter(_.nonEmpty)
+
+                  val afterElipsis: Option[String] =
+                    if (end >= lines.length - 1) None
+                    else Some(elipsis)
+
+                  List(
+                    beforeElipsis,
+                    beforeContext,
+                    Some(line),
+                    Some((1 to caret.col).map(_ => " ").mkString("") + "^"),
+                    Some(errorMsg),
+                    afterContext,
+                    afterElipsis
+                  ).flatten.mkString("\n")
+                }
+              }
+            }
+            case None => "no input"
```

Comment:
can we just use total offsets here and not print nothing? We can still show the `Expectations` right?